File: bindings/python/tests/test_binary.py
"""Python-level tests for the puddin binary parameter functions.

Run with: pytest bindings/python/tests/

Exercises:
- plain numpy inputs (SI) via puddin.binary submodule
- plain numpy inputs (SI) via top-level puddin namespace (backward compat)
- astropy.units.Quantity inputs
- pint.Quantity inputs (if pint is installed)
- JAX inputs (if jax is installed)
- unit mismatch raises an error
- numerical correctness matches Rust reference values
"""

from __future__ import annotations

import math
import pytest
import numpy as np

import puddin
import puddin.binary

MSUN_KG = 1.989e30  # approximate solar mass in kg


# ── helpers ───────────────────────────────────────────────────────────────────

def sol(m):
    """Return mass m (solar masses) as a plain SI float (kg)."""
    return np.atleast_1d(np.float64(m * MSUN_KG))


# ── puddin.binary submodule tests ─────────────────────────────────────────────

class TestBinarySubmodule:
    """Verify that the puddin.binary submodule exposes the same functions."""

    def test_submodule_accessible(self):
        assert hasattr(puddin, "binary")

    def test_total_mass(self):
        result = puddin.binary.total_mass(sol(30), sol(20))
        assert math.isclose(result[0] / MSUN_KG, 50.0, rel_tol=1e-8)

    def test_mass_ratio(self):
        result = puddin.binary.mass_ratio(sol(30), sol(15))
        assert math.isclose(result[0], 0.5, rel_tol=1e-8)

    def test_symmetric_mass_ratio(self):
        result = puddin.binary.symmetric_mass_ratio(sol(30), sol(30))
        assert math.isclose(result[0], 0.25, rel_tol=1e-8)

    def test_chirp_mass(self):
        mc = puddin.binary.chirp_mass(sol(30), sol(30))[0] / MSUN_KG
        expected = 60.0 * 0.25 ** (3.0 / 5.0)
        assert math.isclose(mc, expected, rel_tol=1e-8)

    def test_masses_from_chirp_mass_q_roundtrip(self):
        mc = puddin.binary.chirp_mass(sol(30), sol(20))
        q = puddin.binary.mass_ratio(sol(30), sol(20))
        (m1, m2) = puddin.binary.masses_from_chirp_mass_q(mc, q)
        assert math.isclose(m1[0] / MSUN_KG, 30.0, rel_tol=1e-8)
        assert math.isclose(m2[0] / MSUN_KG, 20.0, rel_tol=1e-8)

    def test_masses_from_chirp_mass_eta_roundtrip(self):
        mc = puddin.binary.chirp_mass(sol(30), sol(20))
        eta = puddin.binary.symmetric_mass_ratio(sol(30), sol(20))
        (m1, m2) = puddin.binary.masses_from_chirp_mass_eta(mc, eta)
        assert math.isclose(m1[0] / MSUN_KG, 30.0, rel_tol=1e-8)
        assert math.isclose(m2[0] / MSUN_KG, 20.0, rel_tol=1e-8)

    def test_chi_eff(self):
        x = puddin.binary.chi_eff(
            sol(30), sol(30), np.array([0.5]), np.array([0.5]),
            np.array([0.0]), np.array([0.0])
        )[0]
        assert math.isclose(x, 0.5, rel_tol=1e-8)

    def test_chi_p(self):
        x = puddin.binary.chi_p(
            sol(30), sol(30), np.array([1.0]), np.array([0.0]),
            np.array([math.pi / 2]), np.array([0.0])
        )[0]
        assert math.isclose(x, 1.0, rel_tol=1e-8)

    def test_results_match_toplevel(self):
        """puddin.binary.* and puddin.* must return identical results."""
        m1, m2 = sol(35), sol(25)
        assert np.allclose(puddin.binary.chirp_mass(m1, m2), puddin.chirp_mass(m1, m2))
        assert np.allclose(puddin.binary.total_mass(m1, m2), puddin.total_mass(m1, m2))
        assert np.allclose(puddin.binary.mass_ratio(m1, m2), puddin.mass_ratio(m1, m2))


# ── plain numpy tests ─────────────────────────────────────────────────────────

class TestPlainSI:
    def test_total_mass(self):
        result = puddin.total_mass(sol(30), sol(20))
        assert math.isclose(result[0] / MSUN_KG, 50.0, rel_tol=1e-8)

    def test_mass_ratio(self):
        result = puddin.mass_ratio(sol(30), sol(15))
        assert math.isclose(result[0], 0.5, rel_tol=1e-8)

    def test_symmetric_mass_ratio_equal(self):
        result = puddin.symmetric_mass_ratio(sol(30), sol(30))
        assert math.isclose(result[0], 0.25, rel_tol=1e-8)

    def test_symmetric_mass_ratio_upper_bound(self):
        for m1, m2 in [(10, 5), (100, 1), (50, 50)]:
            eta = puddin.symmetric_mass_ratio(sol(m1), sol(m2))[0]
            assert eta <= 0.25 + 1e-10

    def test_chirp_mass_equal_masses(self):
        mc = puddin.chirp_mass(sol(30), sol(30))[0] / MSUN_KG
        expected = 60.0 * 0.25 ** (3.0 / 5.0)
        assert math.isclose(mc, expected, rel_tol=1e-8)

    def test_chirp_mass_le_total(self):
        for m1, m2 in [(30, 30), (30, 10), (100, 1)]:
            mc = puddin.chirp_mass(sol(m1), sol(m2))[0]
            mt = puddin.total_mass(sol(m1), sol(m2))[0]
            assert mc <= mt + 1e-6

    def test_masses_from_chirp_mass_q_roundtrip(self):
        mc = puddin.chirp_mass(sol(30), sol(20))
        q = puddin.mass_ratio(sol(30), sol(20))
        (m1, m2) = puddin.masses_from_chirp_mass_q(mc, q)
        assert math.isclose(m1[0] / MSUN_KG, 30.0, rel_tol=1e-8)
        assert math.isclose(m2[0] / MSUN_KG, 20.0, rel_tol=1e-8)

    def test_masses_from_chirp_mass_q_equal(self):
        mc = puddin.chirp_mass(sol(30), sol(30))
        (m1, m2) = puddin.masses_from_chirp_mass_q(mc, np.array([1.0]))
        assert math.isclose(m1[0] / MSUN_KG, 30.0, rel_tol=1e-8)
        assert math.isclose(m2[0] / MSUN_KG, 30.0, rel_tol=1e-8)

    def test_masses_from_chirp_mass_eta_roundtrip(self):
        mc = puddin.chirp_mass(sol(30), sol(20))
        eta = puddin.symmetric_mass_ratio(sol(30), sol(20))
        (m1, m2) = puddin.masses_from_chirp_mass_eta(mc, eta)
        assert math.isclose(m1[0] / MSUN_KG, 30.0, rel_tol=1e-8)
        assert math.isclose(m2[0] / MSUN_KG, 20.0, rel_tol=1e-8)

    def test_masses_from_chirp_mass_eta_equal(self):
        mc = puddin.chirp_mass(sol(30), sol(30))
        (m1, m2) = puddin.masses_from_chirp_mass_eta(mc, np.array([0.25]))
        assert math.isclose(m1[0] / MSUN_KG, 30.0, rel_tol=1e-8)
        assert math.isclose(m2[0] / MSUN_KG, 30.0, rel_tol=1e-8)

    def test_chi_eff_aligned(self):
        x = puddin.chi_eff(sol(30), sol(30), np.array([0.5]), np.array([0.5]),
                           np.array([0.0]), np.array([0.0]))[0]
        assert math.isclose(x, 0.5, rel_tol=1e-8)

    def test_chi_p_in_plane(self):
        x = puddin.chi_p(sol(30), sol(30), np.array([1.0]), np.array([0.0]),
                         np.array([math.pi / 2]), np.array([0.0]))[0]
        assert math.isclose(x, 1.0, rel_tol=1e-8)


# ── astropy tests ─────────────────────────────────────────────────────────────

class TestAstropy:
    pytest.importorskip("astropy")

    def test_chirp_mass_with_astropy_units(self):
        from astropy import units as u
        mc_kg = puddin.chirp_mass(30 * u.Msun, 30 * u.Msun)[0]
        msun_kg = (1 * u.Msun).to('kg').value
        mc_msun = mc_kg / msun_kg
        expected = 60.0 * 0.25 ** (3.0 / 5.0)
        assert math.isclose(mc_msun, expected, rel_tol=1e-6)

    def test_wrong_unit_raises(self):
        from astropy import units as u
        with pytest.raises(Exception):  # astropy raises UnitConversionError
            puddin.chirp_mass(30 * u.meter, 30 * u.meter)


# ── pint tests ────────────────────────────────────────────────────────────────

class TestPint:
    pytest.importorskip("pint")

    def test_chirp_mass_with_pint_units(self):
        import pint
        ureg = pint.UnitRegistry()
        m1 = 30 * MSUN_KG * ureg.kg
        m2 = 30 * MSUN_KG * ureg.kg
        mc = puddin.chirp_mass(m1, m2)[0] / MSUN_KG
        expected = 60.0 * 0.25 ** (3.0 / 5.0)
        assert math.isclose(mc, expected, rel_tol=1e-4)

    def test_wrong_unit_raises(self):
        import pint
        ureg = pint.UnitRegistry()
        with pytest.raises(Exception):
            puddin.chirp_mass(30 * ureg.meter, 30 * ureg.meter)


# ── JAX tests ─────────────────────────────────────────────────────────────────

class TestJAX:
    jax = pytest.importorskip("jax")

    def test_chirp_mass_jit(self):
        import jax
        import jax.numpy as jnp

        m1 = jnp.array([30.0 * MSUN_KG])
        m2 = jnp.array([30.0 * MSUN_KG])
        mc_fn = jax.jit(puddin.chirp_mass)
        mc = mc_fn(m1, m2)[0] / MSUN_KG
        expected = 60.0 * 0.25 ** (3.0 / 5.0)
        assert math.isclose(float(mc), expected, rel_tol=1e-4)

    def test_chirp_mass_grad(self):
        import jax
        import jax.numpy as jnp

        m1 = jnp.array([30.0 * MSUN_KG])
        m2 = jnp.array([30.0 * MSUN_KG])

        def scalar_mc(m1, m2):
            return puddin.chirp_mass(m1, m2)[0]

        grad = jax.grad(scalar_mc)(m1, m2)
        # dMc/dm1 at equal masses = Mc * (3/(5*m1) - 1/(5*M))
        # = Mc * (3/5 - 1/10) / m1 = Mc * 1/2 / m1
        mc = puddin.chirp_mass(m1, m2)[0]
        expected_grad = mc * (3.0 / (5.0 * m1[0]) - 1.0 / (5.0 * 2.0 * m1[0]))
        assert math.isclose(float(grad[0]), float(expected_grad), rel_tol=1e-5)

File: bindings/r/tests/testthat/test-binary.R
test_that("total_mass returns sum of component masses", {
  expect_equal(total_mass(30 * MSUN, 30 * MSUN), 60 * MSUN, tolerance = 1e-10)
  expect_equal(total_mass(10 * MSUN,  5 * MSUN), 15 * MSUN, tolerance = 1e-10)
})

test_that("mass_ratio is m2/m1", {
  expect_equal(mass_ratio(30 * MSUN, 30 * MSUN), 1.0,  tolerance = 1e-10)
  expect_equal(mass_ratio(30 * MSUN, 10 * MSUN), 1/3,  tolerance = 1e-10)
})

test_that("symmetric_mass_ratio is 0.25 for equal masses", {
  expect_equal(symmetric_mass_ratio(30 * MSUN, 30 * MSUN), 0.25, tolerance = 1e-10)
  expect_lte(symmetric_mass_ratio(30 * MSUN, 10 * MSUN), 0.25)
})

test_that("chirp_mass matches analytic formula for equal masses", {
  m <- 30 * MSUN
  mc_expected <- 2 * m * 0.25^(3/5)
  expect_equal(chirp_mass(m, m), mc_expected, tolerance = 1e-10)
})

test_that("chi_eff is zero for non-spinning binary", {
  m <- 30 * MSUN
  expect_equal(chi_eff(m, m, 0, 0, 0, 0), 0, tolerance = 1e-15)
})

test_that("chi_eff is 0.5 for equal aligned spins a=0.5", {
  m <- 30 * MSUN
  expect_equal(chi_eff(m, m, 0.5, 0.5, 0, 0), 0.5, tolerance = 1e-10)
})

test_that("chi_p is zero for non-spinning binary", {
  m <- 30 * MSUN
  expect_equal(chi_p(m, m, 0, 0, 0, 0), 0, tolerance = 1e-15)
})

test_that("masses_from_chirp_mass_q roundtrip", {
  mc  <- chirp_mass(30 * MSUN, 20 * MSUN)
  q   <- mass_ratio(30 * MSUN, 20 * MSUN)
  res <- masses_from_chirp_mass_q(mc, q)
  expect_equal(res$m1, 30 * MSUN, tolerance = 1e-8)
  expect_equal(res$m2, 20 * MSUN, tolerance = 1e-8)
})

test_that("masses_from_chirp_mass_q equal masses", {
  mc  <- chirp_mass(30 * MSUN, 30 * MSUN)
  res <- masses_from_chirp_mass_q(mc, 1.0)
  expect_equal(res$m1, 30 * MSUN, tolerance = 1e-8)
  expect_equal(res$m2, 30 * MSUN, tolerance = 1e-8)
})

test_that("masses_from_chirp_mass_eta roundtrip", {
  mc  <- chirp_mass(30 * MSUN, 20 * MSUN)
  eta <- symmetric_mass_ratio(30 * MSUN, 20 * MSUN)
  res <- masses_from_chirp_mass_eta(mc, eta)
  expect_equal(res$m1, 30 * MSUN, tolerance = 1e-8)
  expect_equal(res$m2, 20 * MSUN, tolerance = 1e-8)
})

test_that("masses_from_chirp_mass_eta equal masses", {
  mc  <- chirp_mass(30 * MSUN, 30 * MSUN)
  res <- masses_from_chirp_mass_eta(mc, 0.25)
  expect_equal(res$m1, 30 * MSUN, tolerance = 1e-8)
  expect_equal(res$m2, 30 * MSUN, tolerance = 1e-8)
})

test_that("masses_from_chirp_mass_q vectorised", {
  mc  <- chirp_mass(c(30, 15) * MSUN, c(20, 10) * MSUN)
  q   <- mass_ratio(c(30, 15) * MSUN, c(20, 10) * MSUN)
  res <- masses_from_chirp_mass_q(mc, q)
  expect_length(res$m1, 2L)
  expect_length(res$m2, 2L)
})

test_that("functions are vectorised", {
  m1 <- c(30, 20, 10) * MSUN
  m2 <- c(30, 20,  5) * MSUN
  mc <- chirp_mass(m1, m2)
  expect_length(mc, 3L)
  expect_true(all(mc > 0))
})
